applications/GeoMechanicsApplication/custom_strategies/schemes/geomechanics_time_integration_scheme.hpp

```diff
@@ -0,0 +1,360 @@
+    }
+
+    void Initialize(ModelPart& rModelPart) override
+    {
+        KRATOS_TRY
+
+        SetTimeFactors(rModelPart);
+
+        Scheme<TSparseSpace, TDenseSpace>::mSchemeIsInitialized = true;
+
+        KRATOS_CATCH("")
+    }
+
+    void InitializeSolutionStep(ModelPart& rModelPart,
+                                TSystemMatrixType& A,
+                                TSystemVectorType& Dx,
+                                TSystemVectorType& b) override
+    {
+        KRATOS_TRY
+
+        SetTimeFactors(rModelPart);
+
+        const ProcessInfo& r_current_process_info = rModelPart.GetProcessInfo();
+
+        block_for_each(rModelPart.Elements(), [&r_current_process_info, this](Element& rElement) {
+            if (IsActive(rElement))
+                rElement.InitializeSolutionStep(r_current_process_info);
+        });
+
+        block_for_each(rModelPart.Conditions(), [&r_current_process_info,
+                                                 this](Condition& rCondition) {
+            if (IsActive(rCondition))
+                rCondition.InitializeSolutionStep(r_current_process_info);
+        });
+
+        KRATOS_CATCH("")
+    }
+
+    void Predict(ModelPart& rModelPart,
+                 DofsArrayType& rDofSet,
+                 TSystemMatrixType& A,
+                 TSystemVectorType& Dx,
+                 TSystemVectorType& b) override
+    {
+        this->UpdateVariablesDerivatives(rModelPart);
+    }
+
+    void InitializeNonLinIteration(ModelPart& rModelPart,
+                                   TSystemMatrixType& A,
+                                   TSystemVectorType& Dx,
+                                   TSystemVectorType& b) override
+    {
+        KRATOS_TRY
+
+        const ProcessInfo& r_current_process_info = rModelPart.GetProcessInfo();
+
+        block_for_each(rModelPart.Elements(), [&r_current_process_info, this](Element& rElement) {
+            if (IsActive(rElement))
+                rElement.InitializeNonLinearIteration(r_current_process_info);
+        });
+
+        block_for_each(rModelPart.Conditions(), [&r_current_process_info,
+                                                 this](Condition& rCondition) {
+            if (IsActive(rCondition))
+                rCondition.InitializeNonLinearIteration(r_current_process_info);
+        });
+
+        KRATOS_CATCH("")
+    }
+
+    void FinalizeNonLinIteration(ModelPart& rModelPart,
+                                 TSystemMatrixType& A,
+                                 TSystemVectorType& Dx,
+                                 TSystemVectorType& b) override
+    {
+        KRATOS_TRY
+
+        const ProcessInfo& r_current_process_info = rModelPart.GetProcessInfo();
+
+        block_for_each(rModelPart.Elements(), [&r_current_process_info, this](Element& rElement) {
+            if (IsActive(rElement))
+                rElement.FinalizeNonLinearIteration(r_current_process_info);
+        });
```

Comment:
As discussed, there are a few patterns in this file that keep coming back. One of them is to call a given member function for each active element. We could define a (private) helper member function as follows:
```
    template <typename MemFuncPtr>
    void BlockForEachActiveElement(ModelPart& rModelPart,
                                   MemFuncPtr pMemberFunction)
    {
        const auto& r_current_process_info = rModelPart.GetProcessInfo();
        block_for_each(rModelPart.Elements(), [&r_current_process_info, pMemberFunction](auto& rElement) {
           if (IsActive(rElement)) {
               (rElement.*pMemberFunction)(r_current_process_info);
           }
        });
    }
```
(Note that here I have changed `IsActive` from a member function to a free function (i.e. a nonmember function).)

With the above helper in place, we can simplify the original code to:
```suggestion
        BlockForEachActiveElement(rModelPart, &Element::FinalizeNonLinearIteration);
```

In a similar way we can define a helper function for calling a given member function for each active condition. Do you feel that this suggestion helps to improve the readability of the code?

Reply:
Very nice suggestion, implemented!